Index: jack_umidi.c
===================================================================
--- jack_umidi.c	(revision 3364)
+++ jack_umidi.c	(working copy)
@@ -39,6 +39,8 @@
 #include <pthread.h>
 #include <err.h>
 #include <sysexits.h>
+#include <sys/errno.h>
+#include <sys/sysctl.h>
 
 #include <jack/jack.h>
 #include <jack/midiport.h>
@@ -514,7 +516,7 @@
 	int c;
 	unsigned int x;
 	const char *pname;
-	char devname[64];
+	char devname[64] = "\0";
 
 	while ((c = getopt(argc, argv, "kBd:hP:SC:n:")) != -1) {
 		switch (c) {
@@ -573,10 +575,34 @@
 		else
 			pname = write_name;
 	}
-
-	snprintf(devname, sizeof(devname), "%s-%s",
-	    (port_name != NULL) ? port_name : PACKAGE_NAME, pname);
-
+	
+	char *dev_name = (read_name != NULL) ? read_name : write_name;
+	if ((dev_name != NULL) && (port_name == NULL) &&
+			(strncmp(dev_name, "/dev/umidi", 10) == 0)) {
+		char desc_name[32];
+		snprintf(desc_name, sizeof(desc_name), 
+			"dev.uaudio.%d.%%desc", atoi(dev_name + 10));
+		char desc_value[64];
+		size_t desc_size = sizeof(desc_value);
+		if ((sysctlbyname(desc_name, desc_value, &desc_size, NULL, 0) == 0) ||
+				/* even if there's not enough space in the buffer, 
+				   we'll still get some name */
+				(errno == ENOMEM)) {
+			for (c = 0; c < desc_size; c++) {
+				if (desc_value[c] == ',') {
+					desc_size = c;
+					break;
+				}
+			}
+			desc_value[desc_size] = '\0';
+			snprintf(devname, sizeof(devname), "%s", desc_value);
+		}
+	}
+	if (strlen(devname) == 0) {
+		snprintf(devname, sizeof(devname), "%s-%s",
+				(port_name != NULL) ? port_name : PACKAGE_NAME, pname);
+	}
+	
 	jack_client = jack_client_open(devname,
 	    JackNoStartServer, NULL);
 	if (jack_client == NULL) {
